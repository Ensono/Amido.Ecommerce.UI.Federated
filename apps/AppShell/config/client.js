const {dependencies} = require('../package.json')

const getFederationConfig = REMOTES => ({
  name: 'app-shell',
  filename: 'remote-entry.js',
  remotes: REMOTES,
  shared: {
    ...dependencies,
    react: {
      singleton: true,
      eager: true,
      requiredVersion: dependencies.react,
    },
    'react-dom': {
      singleton: true,
      eager: true,
      requiredVersion: dependencies['react-dom'],
    },
    'react-router-dom': {
      singleton: true,
      eager: true,
      requiredVersion: dependencies['react-router-dom'],
    },
    'react-redux': {
      singleton: true,
      eager: true,
      requiredVersion: dependencies['react-redux'],
    },
  },
})

module.exports = {getFederationConfig}
