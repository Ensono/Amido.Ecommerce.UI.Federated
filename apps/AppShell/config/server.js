const {dependencies} = require('../package.json')

const getFederationConfig = REMOTES => ({
  name: 'app-shell',
  filename: 'remote-entry.js',
  library: {type: 'commonjs'},
  remotes: REMOTES,
  shared: {
    react: {
      singleton: true,
      eager: true,
      requiredVersion: dependencies.react,
    },
  },
})

module.exports = {getFederationConfig}
