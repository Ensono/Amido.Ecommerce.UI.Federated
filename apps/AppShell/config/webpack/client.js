// const webpack = require('webpack')
const nodeExternals = require('webpack-node-externals')

// const getClientEnvironment = require('../env')
// const paths = require('../paths')
const getBaseConfig = require('./base')

// const {packageJsonDeps, version} = require('../../package.json')

module.exports = webpackEnv => {
  const baseConfig = getBaseConfig(webpackEnv)

  // const env = getClientEnvironment(paths.publicUrlOrPath.slice(0, -1))

  // const REMOTE_URLS = JSON.parse(env.raw.REACT_APP_REMOTE_URLS)

  // const REMOTES = Object.entries(REMOTE_URLS)
  //   .map(([name, entry]) => ({
  //     [name]: `${entry}/static/container.js`,
  //   }))
  //   .reduce((acc, n) => ({...acc, ...n}), {})

  return {
    ...baseConfig,
    target: 'node',
    externals: [nodeExternals()],
    externalsPresets: {node: true},
    // plugins: baseConfig.plugins.concat([
    //   new webpack.container.ModuleFederationPlugin({
    //     name: 'webpackHost',
    //     filename: 'remote-entry.js',
    //     remotes: REMOTES,
    //     shared: {
    //       react: {
    //         singleton: true,
    //         eager: true,
    //         requiredVersion: packageJsonDeps.react,
    //       },
    //       'react-dom': {
    //         singleton: true,
    //         eager: true,
    //         requiredVersion: packageJsonDeps['react-dom'],
    //       },
    //     },
    //   }),
    // ]),
  }
}
