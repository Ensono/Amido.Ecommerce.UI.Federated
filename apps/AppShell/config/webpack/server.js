const webpack = require('webpack')

const paths = require('../paths')
const getBaseConfig = require('./base')

const nodeExternals = require('webpack-node-externals')

const getClientEnvironment = require('../env')
const packageJsonDeps = require('../../package.json').peerDependencies

module.exports = function (webpackEnv) {
  const baseConfig = getBaseConfig(webpackEnv)

  const env = getClientEnvironment(paths.publicUrlOrPath.slice(0, -1))

  const REMOTE_URLS = JSON.parse(env.raw.REACT_APP_REMOTE_URLS)

  const REMOTES = Object.entries(REMOTE_URLS)
    .map(([name, entry]) => ({
      [name]: `${entry}/static/container.js`,
    }))
    .reduce((acc, n) => ({ ...acc, ...n }), {})

  return {
    ...baseConfig,
    target: 'node',
    entry: paths.appTsx,
    output: {
      ...baseConfig.output,
      path: paths.appDist,
      filename: 'app.js',
      libraryTarget: 'commonjs2',
    },
    externals: [nodeExternals()],
    externalsPresets: { node: true },
    plugins: baseConfig.plugins.concat([
      new webpack.container.ModuleFederationPlugin({
        name: 'appShell',
        filename: 'remote-entry.js',
        remotes: REMOTES,
        shared: {
          react: {
            singleton: true,
            eager: true,
            requiredVersion: packageJsonDeps.react,
          },
        },
      }),
    ]),
  }
}
