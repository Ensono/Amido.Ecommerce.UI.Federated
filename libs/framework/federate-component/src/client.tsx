/* eslint-disable @typescript-eslint/naming-convention */
import {createContext, lazy} from 'react'

import {Modules, ModulesContainer, RemotesContext} from './types'

declare global {
  function __webpack_init_sharing__(shareScope: string): Promise<void>
  const __webpack_share_scopes__: any
}

/**
 * ModuleFederationPlugin loading code (loaded through \<script\> tags from imported components)
 * cached globally as ctx[remote][module]
 *
 * @remarks
 * Read more {@link https://webpack.js.org/concepts/module-federation/}
 *
 * @param ctx - global object for caching resolved components
 * @param remote - scope for caching
 * @param module - exported module name
 * @param shareScope - webpack configured shareScope
 * @returns React Component
 */
export const getClientComponent = (ctx: RemotesContext, remote: string, module: string, shareScope: string) => {
  ctx[remote] = ctx[remote] || createContext({})
  const modules = ctx[remote] as Modules

  let Component = modules[module]
  if (!Component) {
    Component = lazy(async () => {
      const modulesContainer = (window as {[key: string]: any})[remote] as ModulesContainer
      await __webpack_init_sharing__(shareScope)
      if (!modulesContainer) {
        throw new Error(
          `Federated component ${module} from ${remote} not loaded, the app ${remote} might be down or the remote-entry.js relative to this remote is not available`,
        )
      }
      await modulesContainer.init(__webpack_share_scopes__.default)
      const factory = await modulesContainer.get(module)
      return factory()
    })
    modules[module] = Component
  }
  return Component
}
