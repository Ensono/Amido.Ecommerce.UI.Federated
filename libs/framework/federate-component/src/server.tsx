import React, {lazy} from 'react'

import {constants} from '@batman/constants'
import {Logger} from '@batman/core-logger'
import axios from 'axios'
import {Parser, ProcessNodeDefinitions} from 'html-to-react'
import stringify from 'json-stringify-deterministic'

import {Module, RemotesContext} from './types'

/**
 * Consumer of /prerender endpoint generated by libs/framework/middlewares/prerender.
 * prerendered HTML is converted into a React element and client-side remote-entry.js
 * module federation files are inlined for client-side render and hydration
 *
 * @param ctx - global caching object
 * @param remote - caching key
 * @param module - module to fetch from remote
 * @param props - props passed to component to be rendered
 * @param remoteUrl - host url of prerender endpoint
 * @returns React component with required <script> and <style> tags to load on client-side
 */
export const getServerComponent = (
  ctx: RemotesContext,
  remote: string,
  module: string,
  props: {[key: string]: any},
  remoteUrl: string,
) => {
  // We cache based on properties. This allows us to only
  // do one fetch for multiple references of a remote component.
  const id = stringify({remote, module, props})

  let Component = ctx[id] as Module

  if (Component) {
    // component is already in the cache, use that
    return Component
  }

  Component = lazy(async () => {
    // Do the post request to pre-render the federated component
    try {
      const res = await axios(`${remoteUrl}/prerender`, {
        method: 'POST',
        data: {
          module,
          props,
        },
        headers: {
          'content-type': 'application/json',
        },
      })
      let parsedChunks: Array<any>
      const [chunks, html] = res.data.split(constants.SERIALISED_RESPONSE_SEPARATOR)
      try {
        parsedChunks = JSON.parse(chunks)
      } catch (err: any) {
        parsedChunks = []
        Logger.error(err)
      }

      const processNodeDefinitions = new ProcessNodeDefinitions(React)
      const parser = new Parser()

      return {
        default: ({children}: any) => {
          const parseInstructions = [
            {
              shouldProcessNode: (node: any) => {
                // If the pre-rendered component rendered a children placeholder,
                // we will process this ourselves.
                if (node?.type === 'text' && node.data === '\u200Cchildren\u200C') {
                  return true
                }
                return false
              },
              processNode: (_: any, __: any, index: number) => {
                // Instead of retaining the children placeholder, render out
                // the children components. This even allows for recursive
                // federated components!
                return <React.Fragment key={index}>{children}</React.Fragment>
              },
            },
            {
              // Process all other nodes with the lib defaults.
              shouldProcessNode: () => true,
              processNode: processNodeDefinitions.processDefaultNode,
            },
          ]

          // Turn the pre-rendered HTML string into a react element
          // while rendering out the children.
          const reactElement = parser.parseWithInstructions(html, () => true, parseInstructions)

          return (
            <>
              {/* Add style chunks and async script tags for the script chunks. */}
              {parsedChunks.map(chunk =>
                chunk.endsWith('.css') ? (
                  <link key={chunk} rel="stylesheet" href={chunk} />
                ) : (
                  <script key={chunk} defer src={chunk} />
                ),
              )}
              {/* Render the re-constructed react element */}
              {reactElement}
            </>
          )
        },
      }
    } catch (err: any) {
      Logger.error(err.message)
      throw new Error(err.message)
    }
  })
  ctx[id] = Component

  return Component
}
