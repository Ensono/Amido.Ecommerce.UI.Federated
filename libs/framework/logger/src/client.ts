/* eslint-disable import/no-extraneous-dependencies */
/* istanbul ignore file */
/* eslint-disable no-console */
const {LOG_LEVEL = 'warn', NODE_ENV, VSCODE_GIT_ASKPASS_NODE} = process.env

const logLevels: any = Object.freeze({
  info: 1,
  debug: 2,
  warn: 3,
  error: 4,
})

const isDev = NODE_ENV === 'development' || VSCODE_GIT_ASKPASS_NODE

const logger = {
  info: (message: string | object, correlationId: any) => {
    if (isDev || logLevels[LOG_LEVEL] === 1) {
      console.log(correlationId, message)
    }
  },
  debug: (message: string | object, correlationId: any) => {
    if (isDev || logLevels[LOG_LEVEL] <= 2) {
      console.debug(correlationId, message)
    }
  },
  warn: (message: string | object, correlationId: any) => {
    if (isDev || logLevels[LOG_LEVEL] <= 3) {
      console.warn(correlationId, message)
    }
    if (typeof window !== 'undefined' && (window as any).appInsights) {
      ;(window as any).appInsights.trackTrace({message, severityLevel: 2})
    }
  },
  error: (message: string | object, correlationId: any) => {
    if (isDev || logLevels[LOG_LEVEL] <= 4) {
      console.error(correlationId, message)
    }
    if (typeof window !== 'undefined' && (window as any)?.appInsights) {
      ;(window as any).appInsights.trackException({error: new Error(message as string), severityLevel: 3})
    }
  },
}

export const Logger = {
  info(message: string | object, correlationId = 'root') {
    logger.info(message, correlationId)
  },
  debug(message: string | object, correlationId = 'root') {
    logger.debug(message, correlationId)
  },
  warn(message: string | object, correlationId = 'root') {
    logger.warn(message, correlationId)
  },
  error(message: string | object, correlationId = 'root') {
    logger.error(message, correlationId)
  },
  stream: {
    write(message: string) {
      logger.debug(message, 'root')
    },
  },
}
