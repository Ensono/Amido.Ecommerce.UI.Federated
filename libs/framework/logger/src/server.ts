/* eslint-disable import/no-extraneous-dependencies */
/* istanbul ignore file */
/* eslint-disable no-console */
import {defaultClient} from 'applicationinsights'

const {LOG_LEVEL = 'warn', NODE_ENV, VSCODE_GIT_ASKPASS_NODE} = process.env

const logLevels: any = Object.freeze({
  info: 1,
  debug: 2,
  warn: 3,
  error: 4,
})

const isDev = NODE_ENV === 'development' || VSCODE_GIT_ASKPASS_NODE

const logger = {
  info: (message: string | object, correlationId: any) => {
    if (isDev || logLevels[LOG_LEVEL] === 1) {
      console.log(correlationId, message)
    }
  },
  debug: (message: string | object, correlationId: any) => {
    if (isDev || logLevels[LOG_LEVEL] <= 2) {
      console.debug(correlationId, message)
    }
  },
  warn: (message: string, correlationId: any) => {
    if (isDev || logLevels[LOG_LEVEL] <= 3) {
      console.warn(correlationId, message)
    }
    if (defaultClient) {
      defaultClient?.trackTrace({message, severity: 2})
    }
  },
  error: (message: string, correlationId: any) => {
    if (isDev || logLevels[LOG_LEVEL] <= 4) {
      console.error(correlationId, message)
    }
    if (defaultClient) {
      defaultClient?.trackException({exception: new Error(message)})
    }
  },
}

export const Logger = {
  info(message: string | object, correlationId = 'root') {
    logger.info(message, correlationId)
  },
  debug(message: string | object, correlationId = 'root') {
    logger.debug(message, correlationId)
  },
  warn(message: string, correlationId = 'root') {
    logger.warn(message, correlationId)
  },
  error(message: string, correlationId = 'root') {
    logger.error(message, correlationId)
  },
  stream: {
    write(message: string) {
      logger.debug(message, 'root')
    },
  },
}
