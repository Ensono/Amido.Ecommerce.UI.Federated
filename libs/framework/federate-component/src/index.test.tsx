import {render} from '@testing-library/react'
import {renderToString} from 'react-dom/server'

import {getClientComponent} from './client'
import {getServerComponent} from './server'

import {federateComponent} from '.'

const MockComponent = ({env}: {env: string}) => <p>What's up yo, I'm on the {env}</p>

jest.mock('./server', () => ({
  getServerComponent: jest.fn(() => MockComponent),
}))

jest.mock('./client', () => ({
  getClientComponent: jest.fn(() => MockComponent),
}))

const props = {
  loadingFallback: <p>loading</p>,
  errorFallback: <p>error</p>,
}

describe('federateComponent', () => {
  afterEach(() => {
    jest.clearAllMocks()
  })

  describe('on the client', () => {
    it('calls getClientComponent', () => {
      federateComponent('remote', 'module', 'remoteUrl')(props)
      expect(getClientComponent).toBeCalledTimes(1)
    })

    it("doesn't call getServerComponent", () => {
      federateComponent('remote', 'module', 'remoteUrl')(props)
      expect(getServerComponent).not.toBeCalled()
    })

    it('returns a component that renders correctly', () => {
      const Component = federateComponent('remote', 'module', 'remoteUrl')
      const clientProps = {...props, env: 'client'}
      const {container} = render(<Component {...clientProps} />)
      expect(container).toMatchSnapshot()
    })

    it.todo('renders the correct error fallback if required')
  })

  describe('on the server', () => {
    const {window} = global

    beforeAll(() => {
      delete (global as any).window
    })

    afterAll(() => {
      global.window = window
    })

    it('calls getServerComponent', () => {
      federateComponent('remote', 'module', 'remoteUrl')({})
      expect(getServerComponent).toBeCalledTimes(1)
    })

    it("doesn't call getClientComponent", () => {
      federateComponent('remote', 'module', 'remoteUrl')({})
      expect(getClientComponent).not.toBeCalled()
    })

    it('returns a component that renders correctly', () => {
      const Component = federateComponent('remote', 'module', 'remoteUrl')
      const serverProps = {...props, env: 'server'}
      // eslint-disable-next-line testing-library/render-result-naming-convention
      const result = renderToString(<Component {...serverProps} />)
      expect(result).toMatchSnapshot()
    })

    it.todo('renders the correct error fallback if required')
  })
})
