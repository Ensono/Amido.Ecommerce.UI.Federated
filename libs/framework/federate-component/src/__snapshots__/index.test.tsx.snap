// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`federateComponent on the client returns a component that renders correctly 1`] = `
<div>
  <p>
    What's up yo, I'm on the 
    client
  </p>
</div>
`;

exports[`federateComponent on the server returns a component that renders correctly 1`] = `"<!--$--><p>What&#x27;s up yo, I&#x27;m on the <!-- -->server</p><!--/$-->"`;
